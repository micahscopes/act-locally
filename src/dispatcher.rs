use crate::{
    message::{Message, MessageKey, Response},
    types::ActorError,
};

pub trait Dispatcher: Send + Sync + 'static {
    fn message_key(&self, message: &dyn Message) -> Result<MessageKey, ActorError>;

    fn wrap(
        &self,
        message: Box<dyn Message>,
        key: MessageKey,
    ) -> Result<Box<dyn Message>, ActorError>;

    fn unwrap(
        &self,
        message: Box<dyn Response>,
        key: MessageKey,
    ) -> Result<Box<dyn Response>, ActorError>;
}
