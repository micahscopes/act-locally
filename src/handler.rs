use crate::message::{Message, MessageDowncast, Response};
use crate::types::ActorError;
use futures::future::LocalBoxFuture;
use futures::FutureExt;
use std::any::Any;
// use std::cell::RefCell;
// use std::rc::Rc;
use std::future::Future;

use smol::lock::RwLock;
use std::sync::Arc;

pub trait AsyncMutatingFunc<'a, S, C, R, E>: Fn(&'a mut S, C) -> Self::Fut + Send + Sync
where
    S: 'static,
    C: 'static,
    R: 'static,
    E: std::error::Error + Send + Sync + 'static,
{
    type Fut: Future<Output = Result<R, E>> + Send;
}

impl<'a, F, S, C, R, E, Fut> AsyncMutatingFunc<'a, S, C, R, E> for F
where
    F: Fn(&'a mut S, C) -> Fut + Send + Sync,
    S: 'static,
    C: 'static,
    R: 'static,
    E: std::error::Error + Send + Sync + 'static,
    Fut: Future<Output = Result<R, E>> + Send,
{
    type Fut = Fut;
}

pub trait AsyncFunc<'a, S, C, R, E>: Fn(&'a S, C) -> Self::Fut + Send + Sync
where
    S: 'static,
    C: 'static,
    R: 'static,
    E: std::error::Error + Send + Sync + 'static,
{
    type Fut: Future<Output = Result<R, E>> + Send;
}

impl<'a, F, S, C, R, E, Fut> AsyncFunc<'a, S, C, R, E> for F
where
    F: Fn(&'a S, C) -> Fut + Send + Sync,
    S: 'static,
    C: 'static,
    R: 'static,
    E: std::error::Error + Send + Sync + 'static,
    Fut: Future<Output = Result<R, E>> + Send,
{
    type Fut = Fut;
}

type BoxAsyncMutatingFunc<S, C, R> =
    Box<dyn for<'a> Fn(&'a mut S, C) -> LocalBoxFuture<'a, Result<R, ActorError>> + Send + Sync>;

pub(crate) struct AsyncMutatingHandler<S, C, R> {
    func: BoxAsyncMutatingFunc<S, C, R>,
}

impl<S: 'static, C: 'static, R: 'static> AsyncMutatingHandler<S, C, R> {
    pub fn new<F, E>(f: F) -> Self
    where
        F: for<'a> AsyncMutatingFunc<'a, S, C, R, E> + 'static,
        E: std::error::Error + Send + Sync + 'static,
    {
        AsyncMutatingHandler {
            func: Box::new(move |s, c| {
                Box::pin(f(s, c).map(|r| r.map_err(|e| ActorError::CustomError(Box::new(e)))))
            }),
        }
    }
}

type BoxAsyncFunc<S, C, R> =
    Box<dyn for<'a> Fn(&'a S, C) -> LocalBoxFuture<'a, Result<R, ActorError>> + Send + Sync>;

pub(crate) struct AsyncHandler<S, C, R> {
    func: BoxAsyncFunc<S, C, R>,
}

impl<S: 'static, C: 'static, R: 'static> AsyncHandler<S, C, R> {
    pub fn new<F, E>(f: F) -> Self
    where
        F: for<'a> AsyncFunc<'a, S, C, R, E> + 'static,
        E: std::error::Error + Send + Sync + 'static,
    {
        AsyncHandler {
            func: Box::new(move |s, c| {
                Box::pin(f(s, c).map(|r| r.map_err(|e| ActorError::CustomError(Box::new(e)))))
            }),
        }
    }
}

pub trait SyncMutatingFunc<'a, S, C, R, E>: Fn(&'a mut S, C) -> Result<R, E> + Send + Sync
where
    S: 'static,
    C: 'static,
    R: 'static,
    E: std::error::Error + Send + Sync + 'static,
{
}

impl<'a, F, S, C, R, E> SyncMutatingFunc<'a, S, C, R, E> for F
where
    F: Fn(&'a mut S, C) -> Result<R, E> + Send + Sync,
    S: 'static,
    C: 'static,
    R: 'static,
    E: std::error::Error + Send + Sync + 'static,
{
}

pub trait SyncFunc<'a, S, C, R, E>: Fn(&'a S, C) -> Result<R, E> + Send + Sync
where
    S: 'static,
    C: 'static,
    R: 'static,
    E: std::error::Error + Send + Sync + 'static,
{
}

impl<'a, F, S, C, R, E> SyncFunc<'a, S, C, R, E> for F
where
    F: Fn(&'a S, C) -> Result<R, E> + Send + Sync,
    S: 'static,
    C: 'static,
    R: 'static,
    E: std::error::Error + Send + Sync + 'static,
{
}

type BoxSyncMutatingFunc<S, C, R> =
    Box<dyn for<'a> Fn(&'a mut S, C) -> Result<R, ActorError> + Send + Sync>;

pub(crate) struct SyncMutatingHandler<S, C, R> {
    func: BoxSyncMutatingFunc<S, C, R>,
}

impl<S: 'static, C: 'static, R: 'static> SyncMutatingHandler<S, C, R> {
    pub fn new<F, E>(f: F) -> Self
    where
        F: for<'a> SyncMutatingFunc<'a, S, C, R, E> + 'static,
        E: std::error::Error + Send + Sync + 'static,
    {
        SyncMutatingHandler {
            func: Box::new(move |s, c| f(s, c).map_err(|e| ActorError::CustomError(Box::new(e)))),
        }
    }
}

type BoxSyncFunc<S, C, R> = Box<dyn for<'a> Fn(&'a S, C) -> Result<R, ActorError> + Send + Sync>;

pub(crate) struct SyncHandler<S, C, R> {
    func: BoxSyncFunc<S, C, R>,
}

impl<S: 'static, C: 'static, R: 'static> SyncHandler<S, C, R> {
    pub fn new<F, E>(f: F) -> Self
    where
        F: for<'a> SyncFunc<'a, S, C, R, E> + 'static,
        E: std::error::Error + Send + Sync + 'static,
    {
        SyncHandler {
            func: Box::new(move |s, c| f(s, c).map_err(|e| ActorError::CustomError(Box::new(e)))),
        }
    }
}

pub(crate) trait AsyncMessageHandler<S>: Send + Sync {
    fn handle<'a>(
        &'a self,
        state: Arc<RwLock<S>>,
        message: Box<dyn Message>,
    ) -> LocalBoxFuture<'a, Result<Box<dyn Response>, ActorError>>;
}

#[allow(unused)]
pub(crate) trait SyncMessageHandler<S>: Send + Sync {
    fn handle_sync<'a>(
        &'a self,
        state: Arc<RwLock<S>>,
        message: Box<dyn Message>,
    ) -> Result<Box<dyn Any + Send>, ActorError>;
}

impl<S, C, R> AsyncMessageHandler<S> for AsyncMutatingHandler<S, C, R>
where
    S: 'static,
    C: 'static + Send,
    R: 'static + Send + Any,
{
    fn handle<'a>(
        &'a self,
        state: Arc<RwLock<S>>,
        message: Box<dyn Message>,
    ) -> LocalBoxFuture<'a, Result<Box<dyn Response>, ActorError>> {
        if let Ok(params) = message.downcast::<C>() {
            Box::pin(async move {
                let mut state = state.write().await;
                let result = (self.func)(&mut *state, *params);
                result.await.map(|r| Box::new(r) as Box<dyn Response>)
            })
        } else {
            Box::pin(async { Err(ActorError::DispatchError) })
        }
    }
}

impl<S, C, R> AsyncMessageHandler<S> for AsyncHandler<S, C, R>
where
    S: 'static,
    C: 'static + Send,
    R: 'static + Send + Any,
{
    fn handle<'a>(
        &'a self,
        state: Arc<RwLock<S>>,
        message: Box<dyn Message>,
    ) -> LocalBoxFuture<'a, Result<Box<dyn Response>, ActorError>> {
        if let Ok(params) = message.downcast::<C>() {
            Box::pin(async move {
                let state = state.read().await;
                let result = (self.func)(&*state, *params);
                result.await.map(|r| Box::new(r) as Box<dyn Response>)
            })
        } else {
            Box::pin(async { Err(ActorError::DispatchError) })
        }
    }
}

impl<S, C, R> AsyncMessageHandler<S> for SyncMutatingHandler<S, C, R>
where
    S: 'static,
    C: 'static + Send,
    R: 'static + Send + Any,
{
    fn handle<'a>(
        &'a self,
        state: Arc<RwLock<S>>,
        message: Box<dyn Message>,
    ) -> LocalBoxFuture<'a, Result<Box<dyn Response>, ActorError>> {
        if let Ok(params) = message.downcast::<C>() {
            Box::pin(async move {
                let result = (self.func)(&mut *state.write().await, *params);
                result.map(|r| Box::new(r) as Box<dyn Response>)
            })
        } else {
            Box::pin(async { Err(ActorError::DispatchError) })
        }
    }
}

impl<S, C, R> SyncMessageHandler<S> for SyncMutatingHandler<S, C, R>
where
    S: 'static,
    C: 'static + Send,
    R: 'static + Send + Any,
{
    fn handle_sync<'a>(
        &'a self,
        state: Arc<RwLock<S>>,
        message: Box<dyn Message>,
    ) -> Result<Box<dyn Any + Send>, ActorError> {
        if let Ok(params) = message.downcast::<C>() {
            let mut state = state.write_blocking();
            (self.func)(&mut *state, *params).map(|r| Box::new(r) as Box<dyn Any + Send>)
        } else {
            Err(ActorError::DispatchError)
        }
    }
}

impl<S, C, R> AsyncMessageHandler<S> for SyncHandler<S, C, R>
where
    S: 'static,
    C: 'static + Send,
    R: 'static + Send + Any,
{
    fn handle<'a>(
        &'a self,
        state: Arc<RwLock<S>>,
        message: Box<dyn Message>,
    ) -> LocalBoxFuture<'a, Result<Box<dyn Response>, ActorError>> {
        if let Ok(params) = message.downcast::<C>() {
            Box::pin(async move {
                let result = (self.func)(&*state.read().await, *params);
                result.map(|r| Box::new(r) as Box<dyn Response>)
            })
        } else {
            Box::pin(async { Err(ActorError::DispatchError) })
        }
    }
}

impl<S, C, R> SyncMessageHandler<S> for SyncHandler<S, C, R>
where
    S: 'static,
    C: 'static + Send,
    R: 'static + Send + Any,
{
    fn handle_sync<'a>(
        &'a self,
        state: Arc<RwLock<S>>,
        message: Box<dyn Message>,
    ) -> Result<Box<dyn Any + Send>, ActorError> {
        if let Ok(params) = message.downcast::<C>() {
            (self.func)(&*state.read_blocking(), *params)
                .map(|r| Box::new(r) as Box<dyn Any + Send>)
        } else {
            Err(ActorError::DispatchError)
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::message::ResponseDowncast;

    use super::*;
    use futures::executor::block_on;

    #[test]
    fn test_async_non_mutating_message_handler() {
        let state = Arc::new(RwLock::new(0u64));
        async fn handler_fn(s: &u64, c: u64) -> Result<u64, ActorError> {
            Ok(s + c)
        }

        let handler = AsyncHandler::new(handler_fn);

        let message = Box::new(10u64);
        let result =
            block_on(handler.handle(state.clone(), message)).expect("Failed to handle message");

        let result = result.downcast_ref::<u64>().unwrap();
        assert_eq!(*result, 10);
    }

    #[test]
    fn test_async_mutating_message_handler() {
        let state = Arc::new(RwLock::new(0u64));
        async fn handler_fn(s: &mut u64, c: u64) -> Result<u64, ActorError> {
            *s += c;
            Ok(*s)
        }

        let handler = AsyncMutatingHandler::new(handler_fn);

        let message = Box::new(10u64);
        let result =
            block_on(handler.handle(state.clone(), message)).expect("Failed to handle message");
        let result = result.downcast::<u64>().unwrap();
        assert_eq!(*result, 10);
    }

    #[test]
    fn test_sync_non_mutating_message_handler() {
        let state = Arc::new(RwLock::new(0u64));
        fn handler_fn(s: &u64, c: u64) -> Result<u64, ActorError> {
            Ok(s + c)
        }

        let handler = SyncHandler::new(handler_fn);

        let message: Box<dyn Message> = Box::new(10u64);
        let result =
            block_on(handler.handle(state.clone(), message)).expect("Failed to handle message");
        let result = result.downcast::<u64>().unwrap();
        assert_eq!(*result, 10);
    }

    #[test]
    fn test_sync_mutating_message_handler() {
        let state = Arc::new(RwLock::new(0u64));
        fn handler_fn(s: &mut u64, c: u64) -> Result<u64, ActorError> {
            *s += c;
            Ok(*s)
        }

        let handler = SyncMutatingHandler::new(handler_fn);

        let message = Box::new(10u64);
        let result =
            block_on(handler.handle(state.clone(), message)).expect("Failed to handle message");
        let result = result.downcast::<u64>().unwrap();
        assert_eq!(*result, 10);
    }

    #[test]
    fn test_store_handlers_in_hashmap() {
        use std::collections::HashMap;

        let state = Arc::new(RwLock::new(0u64));

        async fn async_handler_mut(s: &mut u64, c: u64) -> Result<u64, ActorError> {
            *s += c;
            Ok(*s)
        }

        async fn async_handler_immut(s: &u64, _c: u64) -> Result<u64, ActorError> {
            Ok(*s)
        }

        fn sync_handler_mut(s: &mut u64, c: u64) -> Result<u64, ActorError> {
            *s += c;
            Ok(*s)
        }

        fn sync_handler_immut(s: &u64, _c: u64) -> Result<u64, ActorError> {
            Ok(*s)
        }

        let mut handlers: HashMap<&str, Box<dyn AsyncMessageHandler<u64>>> = HashMap::new();

        handlers.insert(
            "async_mut",
            Box::new(AsyncMutatingHandler::new(async_handler_mut)),
        );
        handlers.insert(
            "async_immut",
            Box::new(AsyncHandler::new(async_handler_immut)),
        );
        handlers.insert(
            "sync_mut",
            Box::new(SyncMutatingHandler::new(sync_handler_mut)),
        );
        handlers.insert("sync_immut", Box::new(SyncHandler::new(sync_handler_immut)));

        let message = Box::new(10u64);

        let result = block_on(handlers["async_mut"].handle(state.clone(), message))
            .expect("Failed to handle message");
        let result = result.downcast::<u64>().unwrap();
        assert_eq!(*result, 10);

        let message = Box::new(10u64);
        let result = block_on(handlers["async_immut"].handle(state.clone(), message))
            .expect("Failed to handle message");
        let result = result.downcast::<u64>().unwrap();
        assert_eq!(*result, 10);

        let message = Box::new(10u64);
        let result = block_on(handlers["sync_mut"].handle(state.clone(), message))
            .expect("Failed to handle message");
        let result = result.downcast::<u64>().unwrap();
        assert_eq!(*result, 20);

        let message = Box::new(10u64);
        let result = block_on(handlers["sync_immut"].handle(state.clone(), message))
            .expect("Failed to handle message");
        let result = result.downcast::<u64>().unwrap();
        assert_eq!(*result, 20);
    }

    #[test]
    fn test_store_sync_handlers_in_hashmap() {
        use std::collections::HashMap;

        let state = Arc::new(RwLock::new(0u64));

        fn sync_handler_mut(s: &mut u64, c: u64) -> Result<u64, ActorError> {
            *s += c;
            Ok(*s)
        }

        fn sync_handler_immut(s: &u64, c: u64) -> Result<u64, ActorError> {
            Ok(s + c)
        }

        let mut handlers: HashMap<&str, Box<dyn SyncMessageHandler<u64>>> = HashMap::new();

        handlers.insert(
            "sync_mut",
            Box::new(SyncMutatingHandler::new(sync_handler_mut)),
        );
        handlers.insert("sync_immut", Box::new(SyncHandler::new(sync_handler_immut)));

        let message = Box::new(10u64);
        let result = handlers["sync_mut"].handle_sync(state.clone(), message);
        let result = result.unwrap().downcast::<u64>().unwrap();
        assert_eq!(*result, 10);

        let message = Box::new(10u64);
        let result = handlers["sync_immut"].handle_sync(state.clone(), message);
        let result = result.unwrap().downcast::<u64>().unwrap();
        assert_eq!(*result, 20);
    }

    #[test]
    fn test_async_non_mutating_func_wrapper() {
        let state = 0;
        async fn handler_fn(s: &u64, c: u64) -> Result<u64, ActorError> {
            Ok(s + c)
        }

        let handler = AsyncHandler::new(handler_fn);

        let result = block_on((handler.func)(&state, 10));
        assert_eq!(result.unwrap(), 10);
    }

    #[test]
    fn test_async_mutating_func_wrapper() {
        let mut state = 0;
        async fn handler_fn(s: &mut u64, c: u64) -> Result<u64, ActorError> {
            *s += c;
            Ok(*s)
        }

        let handler = AsyncMutatingHandler::new(handler_fn);

        let result = block_on((handler.func)(&mut state, 10));
        assert_eq!(result.unwrap(), 10);
    }

    #[test]
    fn test_sync_non_mutating_func_wrapper() {
        let state = 0;
        fn handler_fn(s: &u64, c: u64) -> Result<u64, ActorError> {
            Ok(s + c)
        }

        let handler = SyncHandler::new(handler_fn);

        let result = (handler.func)(&state, 10);
        assert_eq!(result.unwrap(), 10);
    }

    #[test]
    fn test_sync_mutating_func_wrapper() {
        let mut state = 0;
        fn handler_fn(s: &mut u64, c: u64) -> Result<u64, ActorError> {
            *s += c;
            Ok(*s)
        }

        let handler = SyncMutatingHandler::new(handler_fn);

        let result = (handler.func)(&mut state, 10);
        assert_eq!(result.unwrap(), 10);
    }
}
