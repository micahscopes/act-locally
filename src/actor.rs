use crate::dispatcher::Dispatcher;
use crate::handler::{
    AsyncFunc, AsyncHandler, AsyncMessageHandler, AsyncMutatingFunc, AsyncMutatingHandler,
    SyncFunc, SyncHandler, SyncMutatingFunc, SyncMutatingHandler,
};
use crate::message::{ActorMessage, MessageKey, Response, ResponseDowncast};
use crate::types::ActorError;
use ::futures::channel::mpsc;
// use crate::registration::HandlerRegistration;
// use futures::channel::mpsc;
use smol::future::{self, FutureExt};
use smol::stream::StreamExt;
use smol::LocalExecutor;
use std::cell::RefCell;
use std::collections::HashMap;
use std::fmt::Debug;
use std::panic::{self, AssertUnwindSafe};
use std::rc::Rc;
use std::sync::{Arc, RwLock};
use std::time::Duration;
use tracing::instrument::WithSubscriber;
use tracing::subscriber::{set_default, DefaultGuard};
use tracing::{debug, error, info, info_span, span, Instrument, Level, Span, Subscriber};

type StateRef<S> = Rc<RefCell<S>>;
type HandlerMap<S> = Arc<RwLock<HashMap<MessageKey, Box<dyn AsyncMessageHandler<S>>>>>;

pub struct Actor<S: 'static> {
    state: StateRef<S>,
    handlers: HandlerMap<S>,
    receiver: mpsc::UnboundedReceiver<ActorMessage>,
}

impl<S: 'static> Actor<S> {
    pub fn new(state: S) -> (Self, ActorRef<S>) {
        let handlers = Arc::new(RwLock::new(HashMap::new()));
        let (sender, receiver) = mpsc::unbounded();
        (
            Self {
                state: Rc::new(RefCell::new(state)),
                handlers: handlers.clone(),
                receiver,
            },
            ActorRef {
                sender,
                handlers: handlers.clone(),
            },
        )
    }
    pub fn spawn<F>(init_state: F) -> Result<ActorRef<S>, ActorError>
    where
        F: FnOnce() -> Result<S, ActorError>,
        F: 'static + Send,
    {
        Self::setup_spawn_with_subscriber(init_state, || None)
    }

    pub fn spawn_with_tracing<F, T>(
        init_state: F,
        init_subscriber: T,
    ) -> Result<ActorRef<S>, ActorError>
    where
        F: FnOnce() -> Result<S, ActorError>,
        F: 'static + Send,
        T: FnOnce() -> Option<DefaultGuard>,
        T: 'static + Send,
    {
        Self::setup_spawn_with_subscriber(init_state, init_subscriber)
    }

    fn setup_spawn_with_subscriber<F, T>(
        init_state: F,
        init_subscriber: T,
    ) -> Result<ActorRef<S>, ActorError>
    where
        F: FnOnce() -> Result<S, ActorError>,
        F: 'static + Send,
        T: FnOnce() -> Option<DefaultGuard>,
        T: 'static + Send,
    {
        info!("Spawning actor...");

        // Create a channel to communicate between threads
        let (tx, rx) = std::sync::mpsc::sync_channel(1);
        let actor_parent_span = tracing::info_span!("actor", y = 22);
        let actor_thread_span = span!(parent: &actor_parent_span, Level::INFO, "actor", x = 32);
        let actor_parent_span = actor_parent_span.entered();
        // .child_of(&actor_parent_span);

        // Spawn a new thread for the actor
        info!("Spawning new thread for actor");
        std::thread::spawn(move || {
            {
                let tracing_hook = init_subscriber();
                info!("Building local executor on spawned thread");
                let exec = LocalExecutor::new();
                let actor_task_span = info_span!("actor task", z = 44);
                let task = exec.spawn(
                    async move {
                        // let actor_task_span = actor_task_span.enter();
                        info!("Constructing actor state");
                        match init_state() {
                            Ok(state) => {
                                let (mut actor, actor_ref) = Actor::new(state);
                                tx.send(Ok(actor_ref.clone())).unwrap();
                                info!("Starting actor run loop");
                                let run_result = AssertUnwindSafe(actor.run()).catch_unwind().await;
                                if let Err(panic_info) = run_result {
                                    error!("Actor run loop panicked: {:?}", panic_info);
                                }
                            }
                            Err(e) => {
                                error!("Uh oh... Spawn error: {e:?}");
                                tx.send(Err(e)).unwrap();
                            }
                        };
                        info!("Actor run loop finished");
                        // drop(actor_task_span);
                    }
                    .instrument(actor_task_span),
                );

                future::block_on(exec.run(task).with_current_subscriber());
                drop(tracing_hook);
            }
            .instrument(actor_thread_span)
        });

        drop(actor_parent_span);
        // Receive the ActorRef and Dispatcher from the spawned thread
        rx.recv_timeout(Duration::from_secs(2))
            .map_err(|e| ActorError::SpawnError(Box::new(e)))
            .and_then(|result| match result {
                Ok(actor_ref) => {
                    info!("Received ActorRef and Dispatcher from spawned thread");
                    Ok(actor_ref)
                }
                Err(e) => {
                    error!("Failed to receive ActorRef and Dispatcher: {:?}", e);
                    Err(ActorError::SpawnError(Box::<ActorError>::new(e)))
                }
            })
    }

    pub async fn run(&mut self) -> Result<(), ActorError> {
        let result = panic::AssertUnwindSafe(async {
            info!("Actor run loop started");
            while let Some(message) = self.receiver.next().await {
                info!("Actor got a new message...");
                match message {
                    ActorMessage::Notification(key, payload) => {
                        let handlers = self.handlers.read().unwrap();
                        if let Some(handler) = handlers.get(&key) {
                            info!("Handling notification with key: {key:?}");
                            let mut state = self.state.borrow_mut();
                            if let Err(e) = handler.handle(&mut *state, payload).await {
                                info!("Handler error: {e:?}");
                            }
                        } else {
                            error!("No handler found for key: {key:?}");
                        }
                    }
                    ActorMessage::Request(key, payload, response_tx) => {
                        let handlers = self.handlers.read().unwrap();
                        if let Some(handler) = handlers.get(&key) {
                            info!("Handling request with key: {key:?}");
                            let mut state = self.state.borrow_mut();
                            let result = handler.handle(&mut *state, payload).await;
                            let _ = response_tx.send(result);
                        } else {
                            error!("No handler found for key: {:?}", key);
                            let _ = response_tx.send(Err(ActorError::HandlerNotFound));
                        }
                    }
                }
            }
            info!("Actor run loop finished");
        })
        .catch_unwind()
        .await;

        match result {
            Ok(_) => {
                info!("Actor run loop somehow finished: {:?}", result);
                Ok(())
            }
            Err(panic_info) => {
                error!("Actor run loop panicked: {:?}", panic_info);
                Err(ActorError::HandlerPanicked)
            }
        }
    }
}

pub struct ActorRef<S> {
    sender: mpsc::UnboundedSender<ActorMessage>,
    handlers: HandlerMap<S>,
}

impl<S> Clone for ActorRef<S> {
    fn clone(&self) -> Self {
        ActorRef {
            sender: self.sender.clone(),
            handlers: self.handlers.clone(),
        }
    }
}

impl<S: 'static> ActorRef<S> {
    pub async fn ask<M: Send + 'static, R: Send + Debug + 'static, D: Dispatcher>(
        &self,
        dispatcher: &D,
        message: M,
    ) -> Result<R, ActorError> {
        let key = dispatcher.message_key(&message)?;
        let wrapped = dispatcher.wrap(Box::new(message), key.clone())?;

        let (response_tx, response_rx) = ::futures::channel::oneshot::channel();

        self.sender
            .unbounded_send(ActorMessage::Request(key.clone(), wrapped, response_tx))
            .map_err(|_| ActorError::SendError)?;

        let result = response_rx.await.map_err(|_| ActorError::SendError)??;
        debug!(
            "in our ask handler we unwrapped the result as: {:?}",
            &result.any_response()
        );
        let unwrapped = dispatcher.unwrap(result, key.clone());
        debug!("Unwrapped type: {:?}", unwrapped.any_response().type_id());
        debug!("And unwrapped it to {:?}", &unwrapped.any_response());
        unwrapped.and_then(|r| {
            r.downcast::<R>()
                .map(|r| *r)
                .map_err(|_| ActorError::DowncastError)
        })
    }

    pub fn tell<M: Send + 'static, D: Dispatcher>(
        &self,
        dispatcher: &D,
        message: M,
    ) -> Result<(), ActorError> {
        let key = dispatcher.message_key(&message)?;
        let wrapped = dispatcher.wrap(Box::new(message), key.clone())?;
        self.sender
            .unbounded_send(ActorMessage::Notification(key, wrapped))
            .map_err(|_| ActorError::SendError)
    }

    pub fn register_handler_async<C, R, E, F>(&self, key: MessageKey, handler: F)
    where
        C: 'static + Send,
        R: 'static + Send,
        E: std::error::Error + Send + Sync + 'static,
        F: for<'a> AsyncFunc<'a, S, C, R, E> + 'static,
    {
        let handler = AsyncHandler::new(handler);
        if let Ok(mut handlers) = self.handlers.write() {
            handlers.insert(key, Box::new(handler));
        }
    }

    pub fn register_handler_sync<C, R, E, F>(&self, key: MessageKey, handler: F)
    where
        C: 'static + Send,
        R: 'static + Send,
        E: std::error::Error + Send + Sync + 'static,
        F: for<'a> SyncFunc<'a, S, C, R, E> + 'static,
    {
        let handler = SyncHandler::new(handler);
        if let Ok(mut handlers) = self.handlers.write() {
            handlers.insert(key, Box::new(handler));
        }
    }

    pub fn register_handler_async_mutating<C, R, E, F>(&self, key: MessageKey, handler: F)
    where
        C: 'static + Send,
        R: 'static + Send,
        E: std::error::Error + Send + Sync + 'static,
        F: for<'a> AsyncMutatingFunc<'a, S, C, R, E> + 'static,
    {
        let handler = AsyncMutatingHandler::new(handler);
        if let Ok(mut handlers) = self.handlers.write() {
            handlers.insert(key, Box::new(handler));
        }
    }

    pub fn register_handler_sync_mutating<C, R, E, F>(&self, key: MessageKey, handler: F)
    where
        C: 'static + Send,
        R: 'static + Send,
        E: std::error::Error + Send + Sync + 'static,
        F: for<'a> SyncMutatingFunc<'a, S, C, R, E> + 'static,
    {
        let handler = SyncMutatingHandler::new(handler);
        if let Ok(mut handlers) = self.handlers.write() {
            handlers.insert(key, Box::new(handler));
        }
    }
}

pub struct HandlerRegistration<'a, S: 'static, D: Dispatcher> {
    pub actor_ref: &'a ActorRef<S>,
    pub dispatcher: &'a mut D,
}

// Tests
#[cfg(test)]
mod tests {
    use crate::message::{Message, MessageDowncast, Response};

    use super::*;
    use std::{borrow::BorrowMut, thread::sleep};

    #[derive(Clone)]
    struct TestState {
        counter: i32,
        messages: Vec<String>,
    }

    #[derive(Clone)]
    struct IncrementMessage(i32);

    #[derive(Clone)]
    struct AppendMessage(String);

    #[derive(Clone)]
    struct GetCounterMessage;

    #[derive(Clone)]
    struct GetMessagesMessage;

    #[derive(Clone)]
    struct TestDispatcher;

    impl Dispatcher for TestDispatcher {
        fn message_key(&self, message: &dyn Message) -> Result<MessageKey, ActorError> {
            if message.is::<IncrementMessage>() {
                Ok(MessageKey::new("increment"))
            } else if message.is::<AppendMessage>() {
                Ok(MessageKey::new("append"))
            } else if message.is::<GetCounterMessage>() {
                Ok(MessageKey::new("get_counter"))
            } else if message.is::<GetMessagesMessage>() {
                Ok(MessageKey::new("get_messages"))
            } else {
                Err(ActorError::DispatchError)
            }
        }

        fn wrap(
            &self,
            message: Box<dyn Message>,
            _key: MessageKey,
        ) -> Result<Box<dyn Message>, ActorError> {
            debug!("Wrapping message of type: {:?}", message.type_id());
            Ok(message)
        }

        fn unwrap(
            &self,
            message: Box<dyn Response>,
            _key: MessageKey,
        ) -> Result<Box<dyn Response>, ActorError> {
            debug!("Unwrapping message of type: {:?}", message.type_id());
            Ok(message)
        }
    }

    impl<'a> HandlerRegistration<'a, TestState, TestDispatcher> {
        fn register_handlers(&mut self) {
            let actor_ref = &mut self.actor_ref.borrow_mut();
            // let dispatcher = &self.dispatcher;
            actor_ref
                .register_handler_async_mutating(MessageKey::new("increment"), increment_handler);

            actor_ref.register_handler_async_mutating(MessageKey::new("append"), append_handler);

            actor_ref.register_handler_async_mutating(
                MessageKey::new("get_counter"),
                get_counter_handler,
            );
            actor_ref.register_handler_async_mutating(
                MessageKey::new("get_messages"),
                get_messages_handler,
            );
        }
    }

    async fn increment_handler(
        state: &mut TestState,
        msg: IncrementMessage,
    ) -> Result<(), ActorError> {
        state.counter += msg.0;
        Ok(())
    }

    async fn append_handler(state: &mut TestState, msg: AppendMessage) -> Result<(), ActorError> {
        state.messages.push(msg.0);
        Ok(())
    }

    async fn get_counter_handler(
        state: &mut TestState,
        _: GetCounterMessage,
    ) -> Result<i32, ActorError> {
        Ok(state.counter)
    }

    async fn get_messages_handler(
        state: &mut TestState,
        _: GetMessagesMessage,
    ) -> Result<Vec<String>, ActorError> {
        Ok(state.messages.clone())
    }

    #[test]
    fn test_actor_with_custom_dispatcher() {
        let mut dispatcher = TestDispatcher;
        let actor_ref = Actor::spawn(|| {
            Ok(TestState {
                counter: 0,
                messages: Vec::new(),
            })
        })
        .expect("Failed to spawn actor");

        HandlerRegistration {
            actor_ref: &mut actor_ref.clone(),
            dispatcher: &mut dispatcher,
        }
        .register_handlers();

        // Test increment
        actor_ref
            .tell(&dispatcher, IncrementMessage(5))
            .expect("Failed to send increment 5");
        actor_ref
            .tell(&dispatcher, IncrementMessage(3))
            .expect("Failed to send increment 3");

        // Test append
        actor_ref
            .tell(&dispatcher, AppendMessage("Hello".to_string()))
            .expect("Failed to send append Hello");
        actor_ref
            .tell(&dispatcher, AppendMessage("World".to_string()))
            .expect("Failed to send append World");

        // Allow some time for processing
        sleep(Duration::from_millis(100));

        // Test get_counter
        future::block_on(async move {
            let counter: i32 = actor_ref
                .ask(&dispatcher, GetCounterMessage)
                .await
                .unwrap_or_else(|e| {
                    println!("Error getting counter: {:?}", e);
                    panic!("Failed to get counter: {:?}", e);
                });
            assert_eq!(counter, 8);

            let messages: Vec<String> = actor_ref
                .ask(&dispatcher, GetMessagesMessage)
                .await
                .expect("Failed to get messages");
            assert_eq!(messages, vec!["Hello".to_string(), "World".to_string()]);
        });

        // future::block_on(task);

        println!("All tests passed successfully!");
    }
}
