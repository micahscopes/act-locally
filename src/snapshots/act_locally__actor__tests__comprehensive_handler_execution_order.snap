---
source: src/actor.rs
expression: "final_order.join(\"\\n\")"
---
(0, 0, AsyncNonMut, start)
(1, 1, SyncNonMut, start)
(1, 1, SyncNonMut, end)
(2, 2, AsyncNonMut, start)
(0, 0, AsyncNonMut, end)
(2, 2, AsyncNonMut, end)
(3, 3, AsyncMut, start)
(3, 3, AsyncMut, end)
(4, 4, SyncNonMut, start)
(4, 4, SyncNonMut, end)
(5, 6, SyncMut, start)
(5, 6, SyncMut, end)
(6, 5, AsyncNonMut, start)
(6, 5, AsyncNonMut, end)
(7, 9, AsyncMut, start)
(7, 9, AsyncMut, end)
(8, 11, SyncNonMut, start)
(8, 11, SyncNonMut, end)
(9, 7, AsyncNonMut, start)
(10, 8, SyncNonMut, start)
(10, 8, SyncNonMut, end)
(9, 7, AsyncNonMut, end)
(11, 10, AsyncNonMut, start)
(11, 10, AsyncNonMut, end)
